🌐fix untranslatable donation field string

cathysarisky opened the pull request:
closes #21766 

This PR adds the "Add a personal note" string to portal's API request when someone tries to make a donation.  And adjusts the API endpoint to take it.

Request: Please review my handling of what could be user-generated content at the api endpoint.  [It shouldn't be, but...]  Sufficient handling, or more needed?

And tests... always tests.

cathysarisky committed:
fix untranslatable donation field string

cathysarisky committed:
Merge branch 'main' into stripe-donation-i18n

cathysarisky committed:
dangit lint

cathysarisky committed:
Merge branch 'stripe-donation-i18n' of https://github.com/cathysarisky/Ghost into stripe-donation-i18n

cathysarisky commented on ghost/i18n/locales/lt/portal.json:
```diff
@@ -15,6 +15,7 @@
     "Account": "Paskyra",
     "Account details updated successfully": "",
```
Hey @coderabbitai, thanks for flagging the accidental deletion of translations.  I'll fix that.  Good bunny.

cathysarisky replied:
Hey @coderabbitai , I'm confused. I don't see those translations in the original, so I don't think I deleted them??

cathysarisky commented on ghost/i18n/locales/pt-BR/portal.json:
```diff
@@ -15,6 +15,7 @@
     "Account": "Conta",
     "Account details updated successfully": "Detalhes da conta atualizados com sucesso",
     "Account settings": "Configurações de conta",
+    "Add a personal note": "",
```
No thank you @coderabbitai.  When a new key is added, the value will usually be added later in a separate PR.

cathysarisky commented on ghost/i18n/locales/sl/portal.json:
```diff
@@ -15,6 +15,8 @@
     "Account details updated successfully": "Podatki o računu so bili uspešno posodobljeni",
     "Account settings": "Nastavitve računa",
+    "Add a personal note": "",
+    "After a free trial ends, you will be charged the regular price for the tier you've chosen. You can always cancel before then.": "Po koncu brezplačnega preizkusa se vam zaračuna redna cena za izbrano stopnjo. Pred tem jo lahko vedno prekličete.",
     "After a free trial ends, you will be charged the regular price for the tier you've chosen. You can always cancel before then.": "Po koncu brezplačnega preizkusa se vam zaračuna redna cena za izbran paket. Pred tem jo lahko vedno prekličete.",
```
Good @coderabbit. You're right, this is an error.

cathysarisky committed:
fix accidentally duplicated SL key, minor lint, add validation.

cathysarisky committed:
lint

cathysarisky commented on ghost/members-api/lib/controllers/RouterController.js:
```diff
@@ -478,6 +478,13 @@ module.exports = class RouterController {
+                })
+            }
+            options.personalNoteText = rawText ?? 'Add a personal note';
             response = await this._createDonationCheckoutSession(options);
```
This text should not be html.  It's just plaintext, possibly with non-Latin characters.

cathysarisky committed:
validation broke the endpoint, oops.

cathysarisky committed:
grrrr

cathysarisky committed:
Merge branch 'main' into stripe-donation-i18n

cathysarisky committed:
Merge branch 'main' into stripe-donation-i18n

sagzy commented on ghost/members-api/lib/controllers/RouterController.js:
```diff
@@ -478,6 +478,13 @@ module.exports = class RouterController {
+                    message: 'Invalid personalNoteText, possibly an i18n error.'
+                });
+            }
+            options.personalNoteText = rawText ?? 'Add a personal note';
```
A couple of notes here:
1. Given that we're reading input from the request body, I think we should also sanitize it against malicious content, with a library like `sanitize-html`
2. I'm not sure we should block the checkout flow if the personal note string is invalid; this sounds very secondary, so I'd rather fallback to English instead of throwing an error

Given those, I'd suggest something among the lines of:
```
function parsePersonalNote(input) {
    if (typeof input !== 'string') {
        return '';
    }

    if (input.length > 255) {
        return '';
    }
  
    const safeInput = sanitizeHtml(input, {
        allowedTags: [],
        allowedAttributes: {}
    });

    return safeInput; 
}
```

and then use it as such in the calling method:
```
options.personalNoteText = parsePersonalNote(req.body.personalNoteText)
```

sagzy replied:
Bonus: I think `personalNote` as name is as readable as `personalNoteText`, so would simplify

sagzy reviewed (commented):
Hey @cathysarisky, thanks for the PR! Left two suggestions in the code. 

As we're changing the API, a e2e test at the API level would highly be appreciated, e.g. in `ghost/core/test/e2e-api/members/donation-checkout-session.test.js`

cathysarisky committed:
Merge branch 'main' into stripe-donation-i18n

cathysarisky commented:
adding a note to myself. I'm going to add a test to the i18n package also for too-long strings, since Stripe won't like them.

cathysarisky committed:
incorporate review - rename param, add sanitize-html

cathysarisky committed:
tests

cathysarisky committed:
add html stripping test

cathysarisky commented:
Hi @sagzy 

I think I've incorporated your feedback.
I added testing in unit tests, to confirm that RouterController is behaving as expected with a variety of donationNote inputs.  Those are in ghost/members-api/test/unit/lib/controllers/router.test.js.  I'm happy with these.

 I'm struggling with the e2e test you requested.   membersAgent is mocked to just return a url for that create-stripe-checkout endpoint.  Other tests (I looked at create-stripe-checkout-session.test.js too) look at outcomes, but this personalNote only appears in the stripe popover.  It has no impact on the payment, email, or anything else that I can test against.

TL;DR: I could use a hint, if you don't feel that the new unit tests are sufficient. :slightly_smiling_face:

cathysarisky commented on ghost/core/test/e2e-api/members/donation-checkout-session.test.js:
```diff
@@ -329,4 +329,39 @@ describe('Create Stripe Checkout Session for Donations', function () {
 
         latestStripePrice.get('nickname').should.have.length(250);
     });
+    it('Can create a checkout session with a personal note included', async function () {
```
This test is worthless.  It's always going to pass because the body is just the returned url.  I need a clue, please. :)

sagzy replied:
Hey @cathysarisky! 

I think you're right about testing the different parsing scenarios at the unit level, here I think the only thing we'd like to test is that we can pass in a `personalNote` param and that doesn't break the checkout behaviour, end-to-end. 

Curious why we're passing the personal note under `custom_fields` in the test? Aren't we passing it as a top-level `personalNote` in the code?

cathysarisky replied:
Good catch, thank you.  I think I got confused between content of that field and the label. *blush*

cathysarisky committed:
improve logging, fix a test

cathysarisky committed:
add a test for bad string length

cathysarisky committed:
Merge branch 'main' into stripe-donation-i18n

cathysarisky committed:
dangit lint
